core: Add support for passing ptexconv args via file a la grit

jonko0493 opened the pull request:
This PR adds support for passing arbitrary args to ptexconv from a `.ptxc` file placed alongside the png or jpg image. A new `add_ptexconv` method has been added which behaves identically to `add_ptexconv_tex4x4` except that it looks for the aforementioned `.ptxc` file and reads the arguments to pass to ptexconv from it. Additionally, it appropriately conditionalizes the output of the `.idx` and `.pal` files based on the format specified in the args.

jonko0493 committed:
core: Add support for passing ptexconv args via file a la grit

AntonioND commented on architectds/architectds.py:
```diff
@@ -1052,6 +1052,57 @@ def add_ptexconv_tex4x4(self, in_dirs, out_dir='ptexconv'):
+            in_path_png = in_out_file.in_path
+            in_args = ''
+            with open(f'{os.path.splitext(in_out_file.in_path)[0]}.ptxc', 'r') as arg_file:
+                in_args = arg_file.readline()
```
So the only thing I'd change is to support comments in ptxc files like in the case of grit files.

jonko0493 replied:
Yes, that makes sense. Args should still be a one liner, so a while statement reading until we find a line that doesn't start with `#` when stripped should do the trick.

AntonioND commented on architectds/architectds.py:
```diff
@@ -1568,6 +1619,57 @@ def add_ptexconv_tex4x4(self, in_dirs, out_dir='ptexconv'):
+            in_path_png = in_out_file.in_path
+            in_args = ''
+            with open(f'{os.path.splitext(in_out_file.in_path)[0]}.ptxc', 'r') as arg_file:
+                in_args = arg_file.readline()
```
Same. Supporting comments would be great.

jonko0493 commented:
@AntonioND from experimentation, ptexconv does not output the palette when `-fp` (manual palette specification) is added. I've made it so that this flag causes the palette to not be added to the build outputs, so instead users should be responsible for adding the palette themselves via `add_files_unchanged`. Do you think this should be documented in the functions themselves or elsewhere?

AntonioND commented:
> @AntonioND from experimentation, ptexconv does not output the palette when `-fp` (manual palette specification) is added. I've made it so that this flag causes the palette to not be added to the build outputs, so instead users should be responsible for adding the palette themselves via `add_files_unchanged`. Do you think this should be documented in the functions themselves or elsewhere?

Hmmm that's a fair thing to document, I agree. Add it to the functions, and maybe you could also add it to the readme, because this is a weird situation. In "Usage guide", right at the end, add a paragraph with "Note about ptexconv" or similar.

Are you sure this is an expected behaviour of ptexconv? It may be a good idea to report it in that repository.

Oh, and remember to rebase and squash your commits after the changes! This should just be one commit in the git history of the tool.

jonko0493 commented:
Yeah, I'll check with garhoogin and see if that's expected behavior. And yup, will rebase when it's ready!

jonko0493 commented:
Per Garhoogin:

> The idea [is that] if you're generating with a fixed palette then outputting it again is redundant.

Makes sense from a tooling perspective. Thinking on it more, is there a non-hacky way I could add a ninja build rule to automatically copy the palette to the output directory with the appropriate output name? I feel like regex-parsing the command line args for exactly one case is a little bit dicey, but would be interested in your thoughts.

AntonioND commented:
There is a rule that will just copy a file, you can use it:

https://github.com/AntonioND/architectds/blob/e252d7b4c47ca8c46aa258e6c7a89b289906a36f/architectds/architectds.py#L1897

Regarding detecting the issue, not sure...

jonko0493 commented:
I'm also not sure. I might just talk to garhoogin and see if he'd be willing to add an argument that copies the palette to the destination, which would probably be the best solution. Then we'd just add a usage note that says "don't use `-fp`, use `-fpc`" or something to that effect.

jonko0493 commented:
Filed https://github.com/Garhoogin/ptexconv/issues/5. I think let's wait for that to merge before merging this.

AntonioND commented:
Well, it would be great if that option was added. In fact, I have modified `ndstool` to accept multiple input directories for nitrofs, and `grit` a bit to allow the user to set the output directory. So yeah, you wouldn't be the first one to modify the tools to help the build system a bit.

AntonioND commented:
What happened to this in the end?

AntonioND commented:
@jonko0493 do you think we should still wait to merge this?

jonko0493 commented:
@AntonioND sorry I missed the first ping! Yeah, I think we should wait for garhoogin. I'll ping the issue in ptexconv.

AntonioND commented:
This should work, right? https://github.com/Garhoogin/ptexconv/releases/tag/v1.5.1.0

AntonioND commented:
I've upgraded the version of ptexconv in pacman in case you need it to finish the PR: https://github.com/blocksds/packages/actions/runs/16247885525

jonko0493 commented:
Awesome! Will get this wrapped up in the next couple of days -- sorry, have been a bit busy!

jonko0493 commented:
@AntonioND I have updated the readme.rst to contain a special note about passing `-fpo` when using ptexconv and also updated the description of the examples in that document. I have also force-pushed the branch, so I believe this is ready to merge at your convenience!

AntonioND commented:
Thanks!
